### PR TITLE
Recipe books

### PR DESCRIPTION
**Pull request by TheShuEd**
## About the PR
adds a system of forgotten recipes. To craft such things, you must first study a special book that unlocks these recipes.

- [x] storing learned recipes in the mind
- [x] learning new recipes
- [x] prevents from crafting unlearned items.
- [ ] a small portion of content in the form of new crafting recipes, and books that teach these recipes.

## Why / Balance
PR as part of the system for the future librarian rework
https://github.com/space-wizards/docs/pull/173

## Media

https://github.com/space-wizards/space-station-14/assets/96445749/3d5b9ab9-2e95-4c00-90e9-57e70b253bf0

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
later
:cl:
- add: Added fun!

### CODE REVIEW

**Comment by notquitehadouken** [comment]
tiders when they can't craft 8000 bats roundstart:

**Comment by K-Dynamic** [comment]
Time to raid the library like it's Project Zomboid :trollface:

**Comment by Hanzdegloker** [comment]
love this, but I fear books being stolen or hoarded, perhaps a printing press machine that is librarian locked that can print copies for a price? A cheap price but a price. late join librarian could mean a RANSACKED library so yeah, also book loaning and just filling up spares would be nice using the press.

**Comment by TheShuEd** [comment]
I think that all libraries will have to be rebuilt, perhaps blocking half of the bookcases behind librarian access.  So he can have more personal space to work with books

**Comment by K-Dynamic** [comment]
> love this, but I fear books being stolen or hoarded, perhaps a printing press machine that is librarian locked that can print copies for a price? A cheap price but a price. late join librarian could mean a RANSACKED library so yeah, also book loaning and just filling up spares would be nice using the press.

Let librarians print books maybe

**Comment by TheShuEd** [comment]
Sound like a real work for librarian! Hit book thiefs, keep the library is clean, and give people recipes

**Comment by thymol-blue** [comment]
Does this work for construction as well (e.g need to read a book to build a girder)? Also can we make the book consumable somehow? This would be fantastic for downstream!

**Comment by Hanzdegloker** [comment]
this kinda open pandora's box of, imagine role starting knowledge like, Engineers round start can make walls and grilles but nobody outside of engie can unless they learn it from the library.

**Comment by thymol-blue** [comment]
Not for SS14 that would be a slippery slope indeed but it would be nice for downstream which is exactly what you described. In CM-13 only certain roles can build stuff (e.g only ComTechs and engineers can build barricades and set up APCs) but your average marine can't.

I was thinking of a way to do this in CM-14 when bam this PR showed up! It is cool if you don't get to it though, just would be nice to have the options and flexibility if you ever get to it. Sorry for sort of derailing this.

File: Content.Client/Construction/ConstructionSystem.cs [hunk 1]
```diff
@@ -169,6 +171,11 @@ private bool HandleUse(in PointerInputCmdHandler.PointerInputCmdArgs args)
             return true;
         }
 
+        public bool IsUserRecipeLeared(EntityUid user, string recipe)
```
Comment (Tayrtahn) [inline Content.Client/Construction/ConstructionSystem.cs:174]:
```suggestion
        public bool IsUserRecipeLearned(EntityUid user, string recipe)
```

File: Content.Shared/Mind/SharedRecipesUnlockSystem.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        if (learned == null)
+            return;
+
+        var user = args.User;
+
+        foreach (var item in recipeTeacher.Comp.Recipes)
+        {
+            args.Verbs.Add(new()
+            {
+                Text = Loc.GetString("paper-recipe-learning-verb-text", ("item",_proto.Index(item).Name)),
```
Comment (Tayrtahn) [inline Content.Shared/Mind/SharedRecipesUnlockSystem.cs:62]:
```suggestion
                Text = Loc.GetString("paper-recipe-learning-verb-text", ("item", _proto.Index(item).Name)),
```

File: Content.Shared/Mind/SharedRecipesUnlockSystem.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    public void AddRecipeToMind(MindLearnedRecipesComponent comp, ProtoId<ConstructionPrototype> recipe)
+    {
+        if (comp.LearnedRecipes.Contains(recipe))
+            return;
+
+        comp.LearnedRecipes.Add(recipe);
+    }
+
+    public bool IsMindRecipeLeared(EntityUid mind, string recipe, MindLearnedRecipesComponent? comp = null)
```
Comment (Tayrtahn) [inline Content.Shared/Mind/SharedRecipesUnlockSystem.cs:113]:
```suggestion
    public bool IsMindRecipeLearned(EntityUid mind, string recipe, MindLearnedRecipesComponent? comp = null)
```

File: Content.Shared/Mind/SharedRecipesUnlockSystem.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        if (comp == null)
+            comp = EnsureComp<MindLearnedRecipesComponent>(mind);
+
+        return comp.LearnedRecipes.Contains(recipe);
+    }
+
+    public bool IsUserRecipeLeared(EntityUid user, string recipe, MindLearnedRecipesComponent? comp = null)
```
Comment (Tayrtahn) [inline Content.Shared/Mind/SharedRecipesUnlockSystem.cs:121]:
```suggestion
    public bool IsUserRecipeLearned(EntityUid user, string recipe, MindLearnedRecipesComponent? comp = null)
```

File: Content.Shared/Mind/SharedRecipesUnlockSystem.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        var mind = _mind.GetMind(user);
+        if (mind == null)
+            return false;
+
+        return IsMindRecipeLeared(mind.Value, recipe);
```
Comment (Tayrtahn) [inline Content.Shared/Mind/SharedRecipesUnlockSystem.cs:127]:
```suggestion
        return IsMindRecipeLearned(mind.Value, recipe);
```

File: Content.Shared/Paper/PaperRecipeTeacher.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+public sealed partial class PaperRecipeTeacherComponent : Component
+{
+    [DataField]
+    public List<ProtoId<ConstructionPrototype>> Recipes = new();
+
+    [DataField]
+    public float DoAfter = 7f;
+
+    [DataField]
+    public LocId PopupText = "paper-recipe-leaning-start";
```
Comment (Tayrtahn) [inline Content.Shared/Paper/PaperRecipeTeacher.cs:22]:
```suggestion
    public LocId PopupText = "paper-recipe-learning-start";
```

File: Resources/Locale/en-US/paper/paper-recipe-learning.ftl [hunk 7]
```diff
@@ -0,0 +1,4 @@
+paper-recipe-learning-verb-text = Study the "{$item}" recipe
+paper-recipe-learning-verb-message = You can peruse this book to learn how to do new things.
+
+paper-recipe-leaning-start = You immerse yourself in the reading, learning how to create a {$item}.
```
Comment (Tayrtahn) [inline Resources/Locale/en-US/paper/paper-recipe-learning.ftl:4]:
```suggestion
paper-recipe-learning-start = You immerse yourself in the reading, learning how to create a {$item}.
```

File: Content.Client/Construction/ConstructionSystem.cs [hunk 8]
```diff
@@ -169,6 +171,11 @@ private bool HandleUse(in PointerInputCmdHandler.PointerInputCmdArgs args)
...
+        {
+            return _recipeUnlock.IsUserRecipeLeared(user, recipe);
```
Comment (Tayrtahn) [inline Content.Client/Construction/ConstructionSystem.cs:176]:
```suggestion
            return _recipeUnlock.IsUserRecipeLearned(user, recipe);
```

File: Content.Server/Construction/ConstructionSystem.Initial.cs [hunk 9]
```diff
@@ -333,6 +335,12 @@ public async Task<bool> TryStartItemConstruction(string prototype, EntityUid use
                 return false;
             }
 
+            if (constructionPrototype.NeedLearn && !_recipeUnlock.IsUserRecipeLeared(user, constructionPrototype.ID))
```
Comment (Tayrtahn) [inline Content.Server/Construction/ConstructionSystem.Initial.cs:338]:
```suggestion
            if (constructionPrototype.NeedLearn && !_recipeUnlock.IsUserRecipeLearned(user, constructionPrototype.ID))
```

File: Content.Server/Construction/ConstructionSystem.Initial.cs [hunk 10]
```diff
@@ -412,6 +420,12 @@ private async void HandleStartStructureConstruction(TryStartStructureConstructio
                 return;
             }
 
+            if (constructionPrototype.NeedLearn && !_recipeUnlock.IsUserRecipeLeared(user, constructionPrototype.ID))
```
Comment (Tayrtahn) [inline Content.Server/Construction/ConstructionSystem.Initial.cs:423]:
```suggestion
            if (constructionPrototype.NeedLearn && !_recipeUnlock.IsUserRecipeLearned(user, constructionPrototype.ID))
```

File: Content.Shared/Construction/Prototypes/ConstructionPrototype.cs [hunk 11]
```diff
@@ -72,6 +72,12 @@ public sealed partial class ConstructionPrototype : IPrototype
     [DataField("entityWhitelist")]
     public EntityWhitelist? EntityWhitelist = null;
 
+    /// <summary>
+    /// Should the player learn the crafting of this item?
```
Comment (Tayrtahn) [inline Content.Shared/Construction/Prototypes/ConstructionPrototype.cs:76]:
```suggestion
    /// Does the player need to learn the recipe to be able to craft this item?
```

**Review by Tayrtahn — CHANGES_REQUESTED** [review]
Just a few typos and a comment tweak.

**Comment by keronshb** [comment]
> this kinda open pandora's box of, imagine role starting knowledge like, Engineers round start can make walls and grilles but nobody outside of engie can unless they learn it from the library.

I would imagine this isn't the use case.  All current crafts wouldn't be included (weapons, walls, etc). The same would go for most future crafts. This is more for specific cases - like an improved maint shotgun or maybe antagonist exclusive construction.

**Comment by baddiePRO** [comment]
> love this, but I fear books being stolen or hoarded, perhaps a printing press machine that is librarian locked that can print copies for a price? A cheap price but a price. late join librarian could mean a RANSACKED library so yeah, also book loaning and just filling up spares would be nice using the press.

give all library books a GPS tracker, and librarian gets a computer to track down any book with a tracker in it?

**Comment by TheShuEd** [comment]
image

File: Resources/Prototypes/Recipes/Construction/forgotten.yml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  needLearn: true
+
+- type: construction
+  name: shock amulet
+  id: AmuletShockProtection
+  graph: MysticAmuletProtection
+  startNode: start
+  targetNode: shock
+  category: construction-category-forgotten
+  description: Crumbs of unknown energies will keep you from electricy.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Recipes/Construction/forgotten.yml:35]:
```suggestion
  description: Crumbs of unknown energies will protect you from electricity.
```

File: Resources/Prototypes/Recipes/Construction/forgotten.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  needLearn: true
+
+- type: construction
+  name: poison amulet
+  id: AmuletPoisonProtection
+  graph: MysticAmuletProtection
+  startNode: start
+  targetNode: poison
+  category: construction-category-forgotten
+  description: Crumbs of unknown energies will keep you from the poisons.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Recipes/Construction/forgotten.yml:47]:
```suggestion
  description: Crumbs of unknown energies will protect you from poison.
```

File: Resources/Prototypes/Recipes/Construction/forgotten.yml [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  needLearn: true
+
+- type: construction
+  name: cold amulet
+  id: AmuletColdProtection
+  graph: MysticAmuletProtection
+  startNode: start
+  targetNode: cold
+  category: construction-category-forgotten
+  description: Crumbs of unknown energies will keep you out of the cosmic cold.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Recipes/Construction/forgotten.yml:23]:
```suggestion
  description: Crumbs of unknown energies will protect you from the cosmic cold.
```

File: Resources/Prototypes/Recipes/Construction/forgotten.yml [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+# "Forgotten" - catergories for things whose crafts are initially unknown to players and they must first learn them through books
+
+- type: construction
+  name: fire amulet
+  id: AmuletFireProtection
+  graph: MysticAmuletProtection
+  startNode: start
+  targetNode: fire
+  category: construction-category-forgotten
+  description: Crumbs of unknown energies will keep you out of the fire.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Recipes/Construction/forgotten.yml:11]:
```suggestion
  description: Crumbs of unknown energies will protect you from fire.
```

File: Resources/Prototypes/Entities/Clothing/Neck/amulets.yml [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+  - type: Construction
+    graph: MysticAmuletProtection
+    node: poison
+
+- type: entity
+  parent: ClothingNeckBase
+  id: ClothingNeckAmuletShockProtection
+  name: mystical amulet
+  suffix: Shock
+  description: Crumbs of unknown energies will keep you from electricy.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Entities/Clothing/Neck/amulets.yml:61]:
```suggestion
  description: Crumbs of unknown energies will protect you from electricity.
```

File: Resources/Prototypes/Entities/Clothing/Neck/amulets.yml [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  - type: Construction
+    graph: MysticAmuletProtection
+    node: cold
+
+- type: entity
+  parent: ClothingNeckBase
+  id: ClothingNeckAmuletPoisonProtection
+  name: mystical amulet
+  suffix: Poison
+  description: Crumbs of unknown energies will keep you from the poisons.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Entities/Clothing/Neck/amulets.yml:44]:
```suggestion
  description: Crumbs of unknown energies will protect you from poison.
```

File: Resources/Prototypes/Entities/Clothing/Neck/amulets.yml [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+  - type: Construction
+    graph: MysticAmuletProtection
+    node: fire
+    
+- type: entity
+  parent: ClothingNeckBase
+  id: ClothingNeckAmuletColdProtection
+  name: mystical amulet
+  suffix: Cold
+  description: Crumbs of unknown energies will keep you out of the cosmic cold.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Entities/Clothing/Neck/amulets.yml:25]:
```suggestion
  description: Crumbs of unknown energies will protect you from the cosmic cold.
```

File: Resources/Prototypes/Entities/Clothing/Neck/amulets.yml [hunk 19]
```diff
@@ -0,0 +1,71 @@
+- type: entity
+  parent: ClothingNeckBase
+  id: ClothingNeckAmuletFireProtection
+  name: mystical amulet
+  description: Crumbs of unknown energies will keep you out of the fire.
```
Comment (Tayrtahn) [inline Resources/Prototypes/Entities/Clothing/Neck/amulets.yml:5]:
```suggestion
  description: Crumbs of unknown energies will protect you from fire.
```

**Review by Tayrtahn — COMMENTED** [review]
For the record, "crumbs" is a somewhat odd word choice (generally associated with food), but I actually kinda like it, since the weirdness fits with the mysticism of these items.

**Comment by TheShuEd** [comment]
maybe later

**Pull request closed without merging**